File: service/src/main/java/eu/tsalliance/tsmedia/service/CategoryService.java
package eu.tsalliance.tsmedia.service;

import eu.tsalliance.tsmedia.exception.NotFoundException;
import eu.tsalliance.tsmedia.exception.ResourceExistsException;
import eu.tsalliance.tsmedia.models.file.FileCategory;
import eu.tsalliance.tsmedia.repository.CategoryRepository;
import eu.tsalliance.tsmedia.validation.Validator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

@Service
public class CategoryService {

    @Autowired
    private CategoryRepository categoryRepository;

    @Autowired
    private Validator validator;

    public Page<FileCategory> list(Pageable pageable) {
        return this.categoryRepository.findAll(pageable);
    }

    public FileCategory getOne(String id) {
        return this.categoryRepository.getOne(id);
    }

    public boolean existsByName(String name) {
        return this.categoryRepository.findByName(name) != null;
    }

    public FileCategory create(FileCategory category) throws Exception {

        this.validator.validateTextAndThrow(category.getName(), "name", true).alpha().minLen(3).maxLen(32).test();

        if(this.existsByName(category.getName())) {
            throw new ResourceExistsException("name", category.getName());
        }

        return this.categoryRepository.saveAndFlush(category);
    }

    public FileCategory update(String id, FileCategory updated) throws Exception {
        FileCategory oldCategory = this.getOne(id);

        // Check if category exists
        if(oldCategory == null) {
            throw new NotFoundException();
        }

        if(this.validator.validateTextAndThrow(updated.getName(), "name", false).alphaNum().minLen(3).maxLen(32).test() && updated.getName() != null) {
            if(this.existsByName(updated.getName())) throw new ResourceExistsException("name", updated.getName());
            oldCategory.setName(updated.getName());
        }

        return this.categoryRepository.saveAndFlush(oldCategory);
    }

    public void delete(String id) {
        this.categoryRepository.deleteById(id);
    }

}

File: api/src/service/categoryService.ts
import { Validator } from "alliance-rest/lib/validation/validator";
import { Page } from "alliance-rest/lib/pagination/page";
import { Pageable } from "alliance-rest/lib/pagination/pageable";
import { NotFoundError } from "alliance-rest/lib/error/notFoundError";

import { RoleRepository } from "../repository/categoryRepository";
import { IRole, Role } from "../models/auth/role";

export class RoleService {

    private static _roleRepository = RoleRepository;
    private static _validator = new Validator();

    /**
     * Create new role
     * @param role Role's data
     * @returns Promise of type Role
     */
    public static async createRole(role: IRole): Promise<Role> {
        let rolenameExists = await Role.findOne({ where: { rolename: role.rolename }});

        this._validator.text("rolename", role.rolename).required().alphaNum().minLen(3).maxLen(24).unique(() => !!rolenameExists).check();
        this._validator.number("hierarchy", role.hierarchy).required().min(0).max(1000).check();
        this._validator.throwErrors();

        return this._roleRepository.create(role);
    }

    /**
     * Update role
     * @param role Role's updated data
     * @returns Promise of type Role
     */
    public static async updateRole(id: string, updated: IRole): Promise<Role> {
        let role: Role = await this.findById(id);
        if(!role) throw new NotFoundError();

        let rolenameExists = await this.existsByRolename(updated.rolename); await Role.findOne({ where: { rolename: role.rolename }});

        this._validator.text("rolename", updated.rolename).alphaNum().minLen(3).maxLen(24).unique(() => !!rolenameExists).check()
        this._validator.number("hierarchy", updated.hierarchy).min(0).max(1000).check();

        if(!updated.permissions) {
            updated.permissions = role.permissions;
        }
        
        this._validator.throwErrors();
        return this._roleRepository.update(id, role);
    }

    /**
     * Delete role by its id
     * @param id Id of role
     */
    public static async deleteMemberById(id: string): Promise<void> {
        this._roleRepository.deleteById(id);
    }

    /**
     * Get a list of roles
     * @param pageable Page settings
     */
    public static async listAll(pageable: Pageable): Promise<Page<Role>> {
        return this._roleRepository.listAll(pageable);
    }

    /**
     * Find a role by its id
     * @param id Role's id
     * @returns Promise of type Role
     */
    public static async findById(id: string): Promise<Role> {
        return this._roleRepository.findById(id);
    }

    /**
     * Find role by its rolename
     * @param rolename Role's name
     * @returns Promise of type Role
     */
    public static async findByRolename(username: string): Promise<Role> {
        return this._roleRepository.findByRolename(username);
    }

    /**
     * Find role by its rolename and create new entry if it does not exist
     * @param rolename Role's rolename
     * @param create Data to be created
     * @returns Promise of type Role
     */
    public static async findByRolenameOrCreate(rolename: string, create: IRole): Promise<Role> {
        return this._roleRepository.findByRolenameOrCreate(rolename, create);
    }

    /**
     * Check if a role with the rolename already exists
     * @param rolename Rolename to check
     * @returns Promise of type boolean
     */
     public static async existsByRolename(rolename: string): Promise<boolean> {
        return this._roleRepository.existsByRolename(rolename);
    }

}